src/lib/nullifier.ts

```diff
@@ -0,0 +1,84 @@
+import {
+  Poseidon,
+  PublicKey,
+  Field,
+  Group,
+  Struct,
+  MerkleMapWitness,
+} from 'snarkyjs';
+import { Nullifier as JsonNullifier } from '../mina-signer/src/TSTypes.js';
+
+export { Nullifier };
+
+/*
+RFC: https://github.com/o1-labs/snarkyjs/issues/756
+Paper: PLUME https://eprint.iacr.org/2022/1255.pdf
+*/
+class Nullifier extends Struct({
+  message: Field,
+  publicKey: PublicKey,
+  public: {
+    nullifier: Group,
+    s: Field,
+  },
+  private: {
+    c: Field,
+    g_r: Group,
+    h_m_pk_r: Group,
+  },
+}) {
+  static fromJSON(json: JsonNullifier): Nullifier {
+    return super.fromJSON(json as any) as Nullifier;
+  }
+
+  /**
+   * Verifies the correctness of the Nullifier. Throws an error if the Nullifier is incorrect.
+   */
+  verify() {
```

Comment:
we definitely need a unit test for `createNullifier()` -> `Nullfier.fromJSON()` -> `Nullifier.verify()`

should also include negative tests (i.e., I can't make `Nullifier.verify()` work without knowing `sk`)